source/common/common/token_bucket_impl.cc

```diff
@@ -25,4 +25,13 @@ bool TokenBucketImpl::consume(uint64_t tokens) {
   return true;
 }
 
+uint64_t TokenBucketImpl::nextTokenAvailableMs() {
+  // If there are tokens available, return immediately.
+  if (tokens_ >= 1) {
+    return 0;
+  }
+  // TODO(ramaraochavali): implement a more precise way that works for very low rate limits.
+  return (1 / fill_rate_) * 1000;
```

Comment:
@ramaraochavali could you explain why the current approach isn't precise and what do you mean by very low rate limits? I'm debugging a problem and this info could be helpful ;)

Reply:
https://github.com/envoyproxy/envoy/pull/4787#discussion_r227001083 - This comment lead to that TODO in code.